internal/controller/controllers/agentclassification_controller.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	aiv1beta1 "github.com/openshift/assisted-service/api/v1beta1"
+	logutil "github.com/openshift/assisted-service/pkg/log"
+	conditionsv1 "github.com/openshift/custom-resource-status/conditions/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/thoas/go-funk"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	AgentClassificationFinalizer = "agentclassification." + aiv1beta1.Group
+	ClassificationLabelPrefix    = "agentclassification." + aiv1beta1.Group + "/"
+)
+
+// AgentClassificationReconciler reconciles a AgentClassification object
+type AgentClassificationReconciler struct {
+	client.Client
+	Log logrus.FieldLogger
+}
+
+//+kubebuilder:rbac:groups=agent-install.openshift.io,resources=agentclassifications,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=agent-install.openshift.io,resources=agentclassifications/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=agent-install.openshift.io,resources=agentclassifications/finalizers,verbs=update
+//+kubebuilder:rbac:groups=agent-install.openshift.io,resources=agents,verbs=get;list;watch
+
+func (r *AgentClassificationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithFields(
+		logrus.Fields{
+			"agent_classification":           req.Name,
+			"agent_classification_namespace": req.Namespace,
+		})
+
+	defer func() {
+		log.Info("AgentClassification Reconcile ended")
+	}()
+
+	log.Info("AgentClassification Reconcile started")
+
+	classification := &aiv1beta1.AgentClassification{}
+	if err := r.Get(ctx, req.NamespacedName, classification); err != nil {
+		log.WithError(err).Errorf("Failed to get AgentClassification %s", req.NamespacedName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Add a finalizer to newly created objects.
```

Comment:
why do you need a finalizer? 
Usually it's used when you need to do some cleanup before the resources is deleted, is that apply to that case?

Reply:
I guess the cleanup would be to delete the labels on the agents in this case?

Reply:
Correct.  The agentLabel controller deletes the label if the agentClassification is being deleted.  The finalizer is used to wait until the label no longer appears before proceeding with the deletion of the agentClassification.